### PR TITLE
doc: Document stable API change procedure

### PR DESCRIPTION
**Pull request by carlescufi**
As discussed in the API meeting, document how a stable API can be changed in a backwards-incompatible manner when required.

### CODE REVIEW

File: doc/development_process/api_lifecycle.rst [hunk 1]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+A stable API, as described above strives to remain backwards-compatible through
+its life-cycle. There are however cases where fulfilling this objective prevents
+technical progress or is simply unfeasible without unreasonable burden on the
+maintenance of the API and its implementation(s).
+
+In order to restrict and control the introduction of a change that breaks the
+promise of backwards compatibility the following steps must be followed whenever
+such a change is considered necessary in order to accept it in the project:
+
+#. An RFC issue must be opened on GitHub with the following content:
```
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]: Don't we have an RFC template that gets automatically populated when you open an issue of that type?

https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=RFC&template=rfc---proposal.md&title=

File: doc/development_process/api_lifecycle.rst [hunk 2]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+   - A title that matches the RFC issue
+   - The actual changes to the API:
+     - Changes to the API header file
+     - Changes to the API implementation(s)
+     - Changes to the relevant API documentation
+     - Changes to Device Tree source and bindings
+   - The changes required to adapt all in-tree users of the API to the change
+   - An entry in the "API Changes" section of the release notes for the next
+     upcoming release
+   - The labels ``API`` and ``API change``
```
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]: and Release Notes

File: doc/development_process/api_lifecycle.rst [hunk 3]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+   RFC issue and the Pull Request
+
+Once the steps above have been completed, the outcome of the proposal will
+depend on the approval of the actual Pull Request by the maintainer of the
+corresponding subystem. As with any other Pull Request, the author can request
+for it to be discussed and ultimately even voted on in any of the regularly
+scheduled meetings.
+
+If the Pull Request is merged then an email must be sent to the mailing lists
+detailed above informing of the change.
```
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]:
```suggestion
detailed above informing them of the change.
```

File: doc/development_process/api_lifecycle.rst [hunk 4]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+      Title:     RFC: API Change: <subsystem>
+      Contents:  - Background information
+                 - Details on the proposed change
+                   - Function call changes
+                   - Device Tree changes (source and bindings)
+                   - Kconfig option changes
+                 - Impact to users of the API, including the steps required
+                   to adapt out-of-tree users of the API to the change
+
+#. A Pull Request must be opened on GitHub that includes the following:
```
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]: I'm confused from the way this is written if the PR has to be opened simultaneously to the RFC. I would expect the RFC to be discussed and agreed upon before the PR is brought out of draft state, as changing all in-tree users is a lot of work.

Comment (henrikbrixandersen) [reply doc/development_process/api_lifecycle.rst]: I am confused as well. Why the requirement for a separate RFC issue _and_ a PR? In my mind, a PR is just an issue with code attached, and since API changes are code changes I think (at least in some cases) they may be better documented/presented as actual code changes as opposed to text in a GitHub issue.

I suggest rewording to allow both an RFC issue and a PR or just a PR.

File: doc/development_process/api_lifecycle.rst (hunk 1 above)
Comment (dbkinder) [reply doc/development_process/api_lifecycle.rst]: Yes, I see that template too, and it is asking for different content than noted below.  I suspect, if agreed to, the template should be updated to reflect this documentation, and the documentation should be updated to say to click on the RFC/Proposal template when submitting the issue.

File: doc/development_process/api_lifecycle.rst [hunk 5]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+
+   - A title that matches the RFC issue
+   - The actual changes to the API:
+     - Changes to the API header file
```
Comment (dbkinder) [inline doc/development_process/api_lifecycle.rst]: need a blank line before the first item in a (sub)list

File: doc/development_process/api_lifecycle.rst (hunk 2 above)
Comment (dbkinder) [inline doc/development_process/api_lifecycle.rst]: there should also be a reference to the Issue number, yes?

File: doc/development_process/api_lifecycle.rst [hunk 6]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+   - The labels ``API`` and ``API change``
+
+#. An email must be sent to the ``tsc``, ``users`` and ``devel`` mailing lists
+   with a subject identical to the RFC issue title and that links to both the
+   RFC issue and the Pull Request
+
+Once the steps above have been completed, the outcome of the proposal will
+depend on the approval of the actual Pull Request by the maintainer of the
+corresponding subystem. As with any other Pull Request, the author can request
+for it to be discussed and ultimately even voted on in any of the regularly
```
Comment (dbkinder) [inline doc/development_process/api_lifecycle.rst]: add a link:
```
in any of the `regularly scheduled meetings
<https://wiki.zephyrproject.org/Zephyr-Committee-and-Working-Group-Meetings>`_.
```

File: doc/releases/release-notes-2.2.rst [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+We are pleased to announce the release of Zephyr kernel version 2.2.0.
+
+Major enhancements with this release include:
+
+* TBD
+
+API Changes
+***********
+
+Deprecated on this release
```
Comment (dbkinder) [inline doc/releases/release-notes-2.2.rst]: ``on this release`` should be ``in this release`` here, and below.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/development_process/api_lifecycle.rst (hunk 1 above)
Comment (nashif) [reply doc/development_process/api_lifecycle.rst]: The RFC template is for general RFCs, we do not want to make it API specific.

File: doc/development_process/api_lifecycle.rst (hunk 4 above)
Comment (nashif) [reply doc/development_process/api_lifecycle.rst]: I would not ask for a PR with an RFC, first RFC, based on the content and discussion, a PR can be opened. Lets not put the cart before the horses here and set expectation that a PR is required to raise a proposal in form of an RFC. 
If someone has some code, sure, they should open a PR, but here we say "a PR must be opened"....

File: doc/development_process/api_lifecycle.rst (hunk 6 above)
Comment (nashif) [reply doc/development_process/api_lifecycle.rst]: Can we just have people send the email to the devel@ please? This is not an announcement, this just someone opening an issue.

**Review by nashif — CHANGES_REQUESTED** [review]

File: doc/development_process/api_lifecycle.rst (hunk 1 above)
Comment (emob-nordic) [reply doc/development_process/api_lifecycle.rst]: I think the idea was to use this template but with extra contents. We can extend existing template or create new - API specific. I would opt for extensions to existing one.

File: doc/development_process/api_lifecycle.rst (hunk 2 above)
Comment (emob-nordic) [reply doc/development_process/api_lifecycle.rst]: (+) all relevant area labels like: Kconfig, DeviceTre, etc

File: doc/development_process/api_lifecycle.rst (hunk 4 above)
Comment (emob-nordic) [reply doc/development_process/api_lifecycle.rst]: I am almost sure that sooner or later someone who opened the RFC issue would be asked to present it in the code. However we can leave opening PR as optional IMO and do not insist on RFC reporter.

File: doc/development_process/api_lifecycle.rst [hunk 8]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+     - Changes to the relevant API documentation
+     - Changes to Device Tree source and bindings
+   - The changes required to adapt all in-tree users of the API to the change
+   - An entry in the "API Changes" section of the release notes for the next
+     upcoming release
+   - The labels ``API`` and ``API change``
+
+#. An email must be sent to the ``tsc``, ``users`` and ``devel`` mailing lists
+   with a subject identical to the RFC issue title and that links to both the
+   RFC issue and the Pull Request
```
Comment (emob-nordic) [inline doc/development_process/api_lifecycle.rst]: If PR exists.

File: doc/development_process/api_lifecycle.rst [hunk 9]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
 - Stable APIs can get backward compatible updates, bug fixes and security fixes
   at any time.
 
+Introducing incompatible changes
```
Comment (emob-nordic) [inline doc/development_process/api_lifecycle.rst]: Shall we somehow present it in the [api lifecycle](https://github.com/zephyrproject-rtos/zephyr/blame/master/doc/development_process/api_lifecycle.rst#L13)?

File: doc/development_process/api_lifecycle.rst [hunk 10]
```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+     - Changes to the API header file
+     - Changes to the API implementation(s)
+     - Changes to the relevant API documentation
+     - Changes to Device Tree source and bindings
+   - The changes required to adapt all in-tree users of the API to the change
+   - An entry in the "API Changes" section of the release notes for the next
+     upcoming release
+   - The labels ``API`` and ``API change``
+
+#. An email must be sent to the ``tsc``, ``users`` and ``devel`` mailing lists
```
Comment (emob-nordic) [inline doc/development_process/api_lifecycle.rst]: Could we put this info in RFC template so community could see it directly from issue reporting side?

**Review by pabigot — COMMENTED** [review]
While I understand and support the desire to clarify what "stable" means and mediate that against "functional", I feeling this is putting too much detail into a process that we don't understand well enough yet.  My perception is based on the deep discussion here regarding whether the first step is an RFC alone, a standalone PR that's self-justifying, or both together; and on how to notify stakeholders.  We're arguing about the construction details of a theoretical bikeshed.  We might instead start with a couple recent and upcoming examples (counter, gpio, eeprom, `usb_enable`, power management) to document and assess how they work(ed), and use that experience to identify an approach.

The [existing RFC process](https://docs.zephyrproject.org/latest/development_process/proposals.html#proposals-and-rfcs) does not discuss content, stakeholder notification, or decision-making.  What's being proposed here involves a specific class of RFCs, so the two really should be aligned.  Perhaps the first step should be to review and extend the generic RFC process.  For example I see that [feature-requests](https://docs.zephyrproject.org/latest/development_process/proposals.html#feature-tracking) are supposed to be reviewed in TSC regularly; should RFCs also default to TSC discussion/approval, but define a way to delegate to a smaller decision-making body?

Stepping back, I don't feel that Zephyr has the level of process maturity and program management necessary for something like this to be effective.  I wish that could change, but I don't see how, and proposing and approving detailed processes has not been a successful path in the past.

Just my opinion.  Carry on.

File: doc/development_process/api_lifecycle.rst (hunk 1 above)
Comment (mbolivar) [reply doc/development_process/api_lifecycle.rst]: The RFC template ought to be able to cover any RFC

**Comment by nashif** [comment]
> The [existing RFC process](https://docs.zephyrproject.org/latest/development_process/proposals.html#proposals-and-rfcs) does not discuss content, stakeholder notification, or decision-making. What's being proposed here involves a specific class of RFCs, so the two really should be aligned.

agree with @pabigot , there should not be a special class of RFCs for APIs, this PR however can set expectations on the content of the RFC if it involves API, I do not think this needs to be aligned with existing process.

File: doc/development_process/api_lifecycle.rst (hunk 6 above)
Comment (aescolar) [inline doc/development_process/api_lifecycle.rst]: > and ultimately even voted on in any of the regularly scheduled meetings.

The only meeting where voting on any significant matter can happen is the TSC. All other groups do not have the authority.

File: doc/development_process/api_lifecycle.rst (hunk 9 above)
Comment (carlescufi) [reply doc/development_process/api_lifecycle.rst]: I don't think so, because the figure is for the API as a whole, and not for selected incompatible changes.

File: doc/development_process/api_lifecycle.rst (hunk 1 above)
Comment (carlescufi) [reply doc/development_process/api_lifecycle.rst]: Fair enough, so what I will do is to adapt the list below to match into the current RFC template we have.

File: doc/development_process/api_lifecycle.rst (hunk 5 above)
Comment (carlescufi) [reply doc/development_process/api_lifecycle.rst]: Not in this case, since this is a code block

File: doc/development_process/api_lifecycle.rst (hunk 4 above)
Comment (carlescufi) [reply doc/development_process/api_lifecycle.rst]: I would vote for enforcing the RFC in every case, just to be able to have a written record of the change in the form of GitHub issue. That will help with release management for example. One however may shorten the RFC contents (detailed description of the changes) if there is a PR accompanying it.

File: doc/development_process/api_lifecycle.rst (hunk 2 above)
Comment (carlescufi) [reply doc/development_process/api_lifecycle.rst]: Indeed! thanks

File: doc/development_process/api_lifecycle.rst (hunk 10 above)
Comment (carlescufi) [reply doc/development_process/api_lifecycle.rst]: No, because this doesn't really apply to all RFCs. But maybe we want that to change in the future, it's a good idea.

**Comment by carlescufi** [comment]
@nashif @mbolivar @henrikbrixandersen @pabigot @emob-nordic @dbkinder Thanks for the review. I've reworked the proposal based on your comments.

**Review by henrikbrixandersen — APPROVED** [review]

File: doc/development_process/api_lifecycle.rst (hunk 5 above)
Comment (mbolivar) [reply doc/development_process/api_lifecycle.rst]: Did you mean to have two colons after following: on line 128 then?

File: doc/development_process/proposals.rst [hunk 11]
```diff
@@ -88,6 +91,17 @@ Some changes do not require an RFC:
 - Addition of new boards, SoCs or drivers to existing subsystems
 - ...
 
+The process in itself consists in creating a GitHub issue with the :ref:`RFC
+label <gh_labels>` that documents the proposal thoroughly. There is and `RFC
```
Comment (mbolivar) [inline doc/development_process/proposals.rst]: s/is and/is an/

File: doc/development_process/proposals.rst [hunk 12]
```diff
@@ -88,6 +91,17 @@ Some changes do not require an RFC:
...
+template`_ included in the main Zephyr GitHub repository that serves as a
+guideline to write a new RFC.
+
+As with Pull Requests, RFCs might require discussion in the context of one of
+the `Zephyr meetings`_ in order to move it forward in cases where there is
+either disagreement or not enough voiced opinions in order to proceed. Make sure
+to either label it appropriately or include it in the corresponding GiHub
```
Comment (mbolivar) [inline doc/development_process/proposals.rst]: GitHub

**Review by mbolivar — CHANGES_REQUESTED** [review]
LGTM, just some typos

**Comment by carlescufi** [comment]
@mbolivar thanks for the second pass. I've addressed the comments now.

File: doc/development_process/api_lifecycle.rst [hunk 13]
```diff
@@ -77,16 +77,92 @@ An API can be declared ``stable`` after fulfilling the following requirements:
...
+- A title that matches the RFC issue
+- A link to the RFC issue
+- The actual changes to the API
+
+  - Changes to the API header file
+  - Changes to the API implementation(s)
+  - Changes to the relevant API documentation
+  - Changes to Device Tree source and bindings
+
+- The changes required to adapt all in-tree users of the API to the change
```
Comment (mnkp) [inline doc/development_process/api_lifecycle.rst]: If the stable API changes refer to driver API are we asking the author of the PR to single-handedly update all the existing drivers and in-tree users of the API?

**Review by mbolivar — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
LGTM, thanks.

File: doc/development_process/api_lifecycle.rst (hunk 13 above)
Comment (carlescufi) [reply doc/development_process/api_lifecycle.rst]: No, you are right. I can amend that

**Comment by carlescufi** [comment]
@mnkp @mbolivar @pabigot @henrikbrixandersen in my [last push](https://github.com/zephyrproject-rtos/zephyr/compare/6355f8c1e6065e5390392a9bac824e50a1907251..af19ba51a2a222e9b7c74993a30b51b8bd0f0d40) I added a paragraph that describes what is an incompatible change.

**Review by nashif — APPROVED** [review]

**Review by mnkp — APPROVED** [review]

**Review by eobalski — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Comment by carlescufi** [comment]
@galak do you want to take a look at this before merging it?

**Review by galak — COMMENTED** [review]
Can you add a .github/ISSUE_TEMPLATE for the RFC STABLE API change?

File: doc/development_process/api_lifecycle.rst [hunk 14]
```diff
@@ -100,6 +183,8 @@ The following are the requirements for deprecating an existing API:
 
   - Mark as deprecated
   - Document the deprecation
+  - Include the deprecation in the  "API Changes" of the release notes for the
```
Comment (galak) [inline doc/development_process/api_lifecycle.rst]: nit: two spaces` in the (space space) "API"`

**Comment by galak** [comment]
Some minor comments, but LGTM.

**Comment by carlescufi** [comment]
> Can you add a .github/ISSUE_TEMPLATE for the RFC STABLE API change?

@galak I actually wanted to avoid this. I mapped the current RFC template sections to the relevant info for this kind of RFC.

**Review by jhedberg — APPROVED** [review]
